src/repo.rs

```diff
@@ -181,80 +271,42 @@ impl DocumentInfo {
     }
 }
 
-/// Signal that the stream or sink on the network adapter is ready to be polled.
-enum WakeSignal {
-    Stream,
-    Sink,
-}
-
-/// Waking mechanism for stream and sinks.
-enum RepoWaker {
-    Stream(Sender<WakeSignal>),
-    Sink(Sender<WakeSignal>),
-}
-
-/// https://docs.rs/futures/latest/futures/task/trait.ArcWake.html
-impl ArcWake for RepoWaker {
-    fn wake_by_ref(arc_self: &Arc<Self>) {
-        let err = match &**arc_self {
-            RepoWaker::Stream(sender) => sender.try_send(WakeSignal::Stream),
-            RepoWaker::Sink(sender) => sender.try_send(WakeSignal::Sink),
-        };
-        if let Err(TrySendError::Disconnected(_)) = err {
-            panic!("Wake sender disconnected.");
-        }
-    }
-}
-
 /// The backend of a repo: the repo runs an event-loop in a background thread.
-pub struct Repo<T: NetworkAdapter> {
+pub struct Repo {
     /// The Id of the repo.
     repo_id: RepoId,
 
-    network_adapter: Box<T>,
     documents: HashMap<DocumentId, DocumentInfo>,
 
-    stream_waker: Arc<RepoWaker>,
-    sink_waker: Arc<RepoWaker>,
-
     /// Messages to send on the network adapter sink.
-    pending_messages: VecDeque<NetworkMessage>,
+    outgoing_messages: VecDeque<NetworkMessage>,
     /// Events received on the network stream, pending processing.
-    pending_events: VecDeque<NetworkEvent>,
+    incoming_messages: VecDeque<NetworkMessage>,
 
     /// Callback on applying sync messages.
     sync_observer: Option<Box<dyn Fn(Vec<DocumentId>) + Send>>,
-
-    /// Receiver of network stream and sink readiness signals.
-    wake_receiver: Receiver<WakeSignal>,
-
     /// Sender and receiver of repo events.
     repo_sender: Option<Sender<RepoEvent>>,
     repo_receiver: Receiver<RepoEvent>,
+
+    outboxes: Arc<Mutex<HashMap<RepoId, Arc<Mutex<Outbox>>>>>,
```

Comment:
Yes I think it's a good to reduce the complexity of the locking, and also that it's not the number of locks in and of itself that increase complexity(the collection of locks here is like one logical concept).

> Everywhere we take a lock on the inner mutex we are already holding a lock on the outer mutex

Yes but the streams only take locks on the inner ones, the outer one is only used outside of the repo when `connect_stream` is called. 

> Maybe it would make more sense to use an RwLock as the outer lock?

Why? I don't see a current case where there would be more than one concurrent reader of it.

Reply:
> Yes but the streams only take locks on the inner ones, the outer one is only used outside of the repo when connect_stream is called.

Gotcha, thanks!

> Why? I don't see a current case where there would be more than one concurrent reader of it.

You're right, nix this idea :slightly_smiling_face: .